Customizing contributors step of deposit workflow for instruments

EricDurante opened the pull request:
I'd like to do more work on this if I have time, but I'm pushing up what I have currently in case I don't.

In terms of UI behavior, I think that this is correct/complete based on what we've discussed, and it should at least convey the gist of how I think that this should work. At the very least, it needs some additional unit tests for some of the new code as it's currently written. Those are missing because I'm not quite settled on the design of the new form objects that I've introduced.

There were a handful of failing tests in the base branch when I started, and those that were directly related to the new instruments pathway are fixed now, but there are a few tests that use VCR that are still failing for me. I'm not sure why they were failing to begin with since they don't seem related to this work, and I haven't looked into it.

In terms of the overall design, I'm also not yet quite happy with how we're handling data validations in general. Right now we have a mixture of validations on the form objects that validate data in the context of the deposit UI steps, and also validations on the ActiveRecord model that validate data in other contexts such as depositing works via the API. This has the effect of layering validations on top of other validations that are sometimes redundant, and it forces us to incorporate conditional logic in some of the validations. Some validations are common/shared, but they're all context-specific to some extent. In the long run, I think that it might be better to remove most or all of the validation logic from the `WorkVersion` ActiveRecord model and validate input from the API using another, separate form/model object that may share some validation logic with the existing UI forms. However, that's a larger refactoring that doesn't specifically have to do with this new deposit pathway.

EricDurante committed:
WIP on customizing contributors step of deposit workflow for instruments

This is a work in progress. At this point, I think that the workflow is
working the way that we want it to, but I haven’t written unit tests for
the new form objects yet because I’m not quite happy with the design,
and I was hoping to improve it a bit first. There are also still some
minor to-dos and potential refactorings to address.

ajkiessl commented:
@EricDurante No problem if you don't have time to finish this.  We'll revisit the instruments stuff in the next ScholarSphere cycle.

EricDurante commented:
Thanks, @ajkiessl. I likely won't have time to finish, but I'm hoping that I can at least push up a little bit more work on it while it's still fresh in my mind.

EricDurante committed:
Refactor some behavior out of the work deposit form base class

The inheritance hierarchy of these form classes is becoming a little
more clear now that there are more of them, and the base class was
collecting too much behavior that wasn’t needed by all of its
subclasses.

ajkiessl committed:
Merge main and resolve conflicts

ajkiessl committed:
niftany

ajkiessl committed:
Ci fix and fixed pathway merge issue

ajkiessl committed:
Fix some tests

ajkiessl committed:
Merge instruments-pathway and resolve conflicts

ajkiessl committed:
Rubocop

ajkiessl committed:
Fix broken tests

ajkiessl committed:
Fixed form bugs

ajkiessl committed:
Move contributors form builder into the deposit pathway class

ajkiessl committed:
Adds unit tests for WorkDepositPathway contributor form classes

ajkiessl committed:
Adds header for instruments contributors

ajkiessl commented:
The 'spec/features/dashboard/work_version_form_spec.rb' instruments and culminating experiences tests had some difficulties merging, but I think it came out still testing what we want it to.

There may be other weird things in here resulting from merge conflicts.

closes #1691

ajkiessl committed:
Fix creator building for collections bug

ajkiessl committed:
Refactored/removed .build_form.  .build_creator already has the logic for when there are already creators.  So we just need the contributors form to override it in the instruments pathway.

Smullz622 commented on app/forms/work_deposit_pathway.rb:
```diff
@@ -363,15 +409,36 @@ def self.form_fields
+          delegate "#{attr_name}=", to: :work_version, prefix: false
+        end
+
+        def build_creator(*)
```
I don't totally get what's going on with this method being empty. Could you please explain?

ajkiessl replied:
Since the `@resource` variable in the contributors controller could be a `Collection`, `ContributorsFormBase`, or `Instrument::ContributorForm` and just calls `.build_creator`, it seemed like the best way to handle the polymorphism without adding logic to the controller.  The `Instrument::ContributorForm` doesn't build creators, so when it calls the method, it doesn't do anything.  It does build creators in the other two instances.

ajkiessl replied:
I updated the comment to make this more clear.

ajkiessl replied:
But I think we still want an associated authorship (creator) for these.  Rethinking this.

ajkiessl replied:
Ah never mind the Work will still have the depositor association

ajkiessl committed:
Adds controller logic back in to prevent creator build if creators exist

ajkiessl committed:
Update no-op comment

ajkiessl committed:
Delegate creators to workversion even though it wont have any

ajkiessl committed:
add these back in

Smullz622 reviewed (approved):
lgtm!
